### ISSUE TITLE
Extending a definition and overriding the concrete may have no effect

### ISSUE THREAD

**Issue by jleeothon** [thephpleague/container#170]
I'm using `league/container=3.2.2`

```php
<?php

require_once 'vendor/autoload.php';

use League\Container\Container;

class MyClass {}
class MyClass2 {}

$container = new Container();

$container->share('a', new MyClass());
var_dump($container->get('a'));
// object(MyClass)#6 (0) {
// }

$container->extend('a')->setConcrete(new MyClass2());
var_dump($container->get('a'));
// Actual:
// object(MyClass)#6 (0) {
// }
// Expected:
// object(MyClass2)#7 (0) {
// }
```

If I understand correctly, this may happen because when updating the `$concrete`, the `$resolved` member is not updated to `null`.

**Comment by odan** [comment]
I'm not sure if it helps, but this "workaround" also works:

```php
$container->add('foo', function () {
    // pseudo shared
    static $value = null;
    if ($value === null) {
        $value = 'bar';
    }

    return $value;
});
var_dump($container->get('foo')); // bar

$container->extend('foo')->setConcrete(function () {
    return 'baz';
});

var_dump($container->get('foo')); // baz
```

**Comment by odan** [comment]
Hello! 

I need this functionality for my unit tests, because I have to overwrite some of the container definitions like the logger, the session and the database connection...

I would be very thankful for merging this fix. :-)

**Comment by philipobenito** [comment]
Fixed by #172
